### PR TITLE
Write Node for X-Drive

### PR DESCRIPTION
**Pull request by Dylan-Powers**
Write a node to control an X-Drive

### CODE REVIEW

File: src/v5_hal/firmware/include/nodes/subsystems/drivetrain_nodes/HolonomicDriveNode.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    void m_setLeftFrontVelocity(int voltage);
+
+    void m_setLeftRearVelocity(int voltage);
+
+    void m_setRightFrontVelocity(int voltage);
+
+    void m_setRightRearVelocity(int voltage);
+
+public: 
+    HolonomicDriveNode(NodeManager* node_manager, std::string handle_name, MotorNode* left_front_motor,
```
Comment (NathanDuPont) [inline src/v5_hal/firmware/include/nodes/subsystems/drivetrain_nodes/HolonomicDriveNode.h:34]: For now we should probably pass in the controller node as well and have teleopPeriodic control the drivetrain (similar to TankDriveNode)

File: src/v5_hal/firmware/src/main.cpp [hunk 2]
```diff
@@ -59,7 +60,9 @@ void initialize() {
 
 	tank_drive_node = new TankDriveNode(node_manager, "drivetrain", primary_controller, 
 		left_front_drive, left_rear_drive, right_front_drive, right_rear_drive);
```
Comment (NathanDuPont) [inline src/v5_hal/firmware/src/main.cpp:62]: Remove this for now and just have the HolonomicDriveNode

**Review by NathanDuPont — CHANGES_REQUESTED** [review]

File: src/v5_hal/firmware/src/main.cpp (hunk 2 above)
Comment (Dylan-Powers) [reply src/v5_hal/firmware/src/main.cpp:62]: The problem with that is that OdometryNode, ProgrammingSkillsAuton, and AutonManagerNode all expect a tank drive so if we removed TankDriveNode we wouldn't be able to push to master

Comment (NathanDuPont) [reply src/v5_hal/firmware/src/main.cpp:62]: ah yeah, that's true. We'll have to refactor those elsewhere, forgot about that. For now I'd probably comment out the x-drive node then (just so we don't have two subsystems sending commands to the same motors) with a comment to fix once the other subsystems are removed.

These changes should be fixed once I eventually get around to having a generic "drivetrain" type

File: src/v5_hal/firmware/src/nodes/subsystems/drivetrain_nodes/HolonomicDriveNode.cpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    m_setLeftRearVelocity(left_rear_velocity / MAX_WHEEL_SPEED * 200);
+    m_setRightFrontVelocity(right_front_velocity / MAX_WHEEL_SPEED * 200);
+    m_setRightRearVelocity(right_rear_velocity / MAX_WHEEL_SPEED * 200);
+}
+
+void HolonomicDriveNode::teleopPeriodic() {
+    m_setLeftFrontVoltage(((m_controller->get_analog(pros::E_CONTROLLER_ANALOG_LEFT_Y)) / 127.0) * MAX_MOTOR_VOLTAGE);
+    m_setLeftRearVoltage(((m_controller->get_analog(pros::E_CONTROLLER_ANALOG_LEFT_Y)) / 127.0) * MAX_MOTOR_VOLTAGE);
+    m_setRightFrontVoltage(((m_controller->get_analog(pros::E_CONTROLLER_ANALOG_LEFT_X)) / 127.0) * MAX_MOTOR_VOLTAGE);
+    m_setRightRearVoltage(((m_controller->get_analog(pros::E_CONTROLLER_ANALOG_LEFT_X)) / 127.0) * MAX_MOTOR_VOLTAGE);
```
Comment (NathanDuPont) [inline src/v5_hal/firmware/src/nodes/subsystems/drivetrain_nodes/HolonomicDriveNode.cpp:91]: The code for controlling the holonomic drive will be a little bit different. This outline from VEX Coding studio will need to be changed around slightly but should have the basics correct: 
image - Let us know if you have any questions about the implementation or the channels being used

Comment (NathanDuPont) [reply src/v5_hal/firmware/src/nodes/subsystems/drivetrain_nodes/HolonomicDriveNode.cpp:91]: The channels are on the controller in red (by the joysticks): 
image

**Review by NathanDuPont — CHANGES_REQUESTED** [review]

**Review by NathanDuPont — APPROVED** [review]

**Pull request merged**